node/actors/network/src/gossip/attestation/mod.rs

```diff
@@ -0,0 +1,296 @@
+        let before = state.weight;
+        let res = state.insert_votes(votes);
+        if state.weight > before {
+            metrics::METRICS.votes_collected.set(state.votes.len());
+            #[allow(clippy::float_arithmetic)]
+            metrics::METRICS
+                .weight_collected
+                .set(state.weight as f64 / state.config.committee.total_weight() as f64);
+            locked.send_replace(Some(state));
+        }
+        res
+    }
+
+    /// All votes kept in the state.
+    pub(crate) fn votes(&self) -> Vec<Arc<attester::Signed<attester::Batch>>> {
+        self.state
+            .subscribe()
+            .borrow()
+            .iter()
+            .flat_map(|s| s.votes.values().cloned())
+            .collect()
+    }
+
+    /// Waits for the certificate for a batch with the given number to be collected.
+    /// Returns None iff attestation already skipped to collecting certificate some later batch.
+    pub async fn wait_for_qc(
+        &self,
+        ctx: &ctx::Ctx,
+        n: attester::BatchNumber,
+    ) -> ctx::OrCanceled<Option<attester::BatchQC>> {
+        let recv = &mut self.state.subscribe();
+        recv.mark_changed();
+        loop {
+            let state = sync::changed(ctx, recv).await?;
+            let Some(state) = state.as_ref() else {
+                continue;
+            };
+            if state.config.batch_to_attest.number < n {
+                continue;
+            };
+            if state.config.batch_to_attest.number > n {
+                return Ok(None);
+            }
+            if let Some(qc) = state.qc() {
+                return Ok(Some(qc));
+            }
+        }
+    }
+
+    /// Updates the attestation config.
+    /// Clears the votes collected for the previous config.
+    /// Batch number has to increase with each update.
+    pub async fn update_config(&self, config: Config) -> anyhow::Result<()> {
+        let locked = self.state.lock().await;
+        let old = locked.borrow().clone();
+        if let Some(old) = old.as_ref() {
+            if *old.config == config {
+                return Ok(());
+            }
+            anyhow::ensure!(
+                old.config.batch_to_attest.genesis == config.batch_to_attest.genesis,
+                "tried to change genesis"
+            );
+            anyhow::ensure!(
+                old.config.batch_to_attest.number < config.batch_to_attest.number,
+                "tried to decrease batch number"
+            );
+        }
+        let mut new = State {
+            config: Arc::new(config),
+            votes: im::HashMap::new(),
+            weight: 0,
+        };
+        if let Some(key) = self.key.as_ref() {
+            if new.config.committee.contains(&key.public()) {
+                let vote = key.sign_msg(new.config.batch_to_attest.clone());
+                // This is our own vote, so it always should be valid.
+                new.insert_vote(Arc::new(vote)).unwrap();
+            }
```

Comment:
As my other comments show I realise that nothing explicitly says the committee and the batch hash are coming from the main node, so _technically_ we're just signing what the configuration says we need to be voting on, which _could_ be coming from our own database. I know we agreed `Client` is a convenience and `Config` or the `StateController` is the actual API.

So my more general observation is that we should document the expectations around the `Config`.

Reply:
afaic it wai: with this PR the semantics of votes gossiping is in fact more restrictive - every node is willing to gossip only votes which are matching the hash that it calculated on its own. It is true that the API has shifted from where it was before, but thanks to that it is smaller. If you want the signing to be done together with verification/hash computation, we can move it out to the zksync-era side (i.e. entirely out of era-consensus), but I disagree that this approach is any more rubber stamping than the previous implementation.

Reply:
I agree with the restrictiveness, I have no problem with it - it required some changes to the protocol to request the resubmission of the votes once the node knows what the hash is, so I had no choice but to implement it the way it was within the constraints of the design at the time. 

I am also not arguing with you about whether the implementation is more rubber stamping than before, simply because I haven't seen the full implementation, only this half of it. My point was that it was unclear from what we see here how the rest of the implementation will evolve. This is for historic reasons and the lack of an updated design documents: we have a "client" and an API on the main node serving some data, and it we haven't discussed how this client could evolve: currently it's called `AttestationStatusClient`; it could evolve into a more general consensus related "main node client", or it could stay special to the attestation like here - I know Bruno wanted to avoid fragmentation, so imagine if we needed 3-4 of these special clients. My _assumption_ was that whatever data it gets it will get from the API, because it's a "client" and we already had `BatchStore::batch_to_sign` - which I see we no longer have. Based on those assumptions it was pure rubber stamping. The previous implementation explicitly retrieved the thing to sign, so it was obviously not rubber stamping. 

IMO if the `Client` serves as a convenience then it should ideally be simple to understand and keep the implementer within the intended guard rails of the architecture. As I mentioned it could do this if we broke up the `config` method into 3 different methods to retrieve: 1) the current status from the main API, then if it changed the 2) batch to vote on once its available and 3) the committee for that batch, which should be available from the previous batch execution - this would also allow us to return `Arc` for the committee if we wanted by caching it. The `Runner` could exercise these methods and pass them atomically to the `update` and it could clearly be explained where each piece of data is expected to come from. Technically it would be the same, but more obvious.

Reply:
To your specific points: 

>  every node is willing to gossip only votes which are matching the hash 

Yes.

> that it calculated on its own.

This wasn't clear from the PR.

> the API has shifted from where it was before, but thanks to that it is smaller.

I assume we consider the API to include the `Client` and the `BatchStore` as well. It is true that it has less methods, which makes it have a smaller footprint, but it's also "bigger" in terms of having more responsibility of implementing undocumented assumptions in how it produces `Config`.

> If you want the signing to be done together with verification/hash computation, we can move it out to the zksync-era side (i.e. entirely out of era-consensus)

No no, I think it's good that it lives here. Perhaps one way to look at it is that if we can verbally explain what should happen, and we _can_ implement the outline of that algorithm in `era-consensus`, then we _should_, with as many of the rules as possible captured in the code, leaving only the holes that we _cannot_ implement in this repo to be filled in on the `zksync-era` side. That would be clear guideline IMO. Alternatively we could start treating it as a pure gossip channel that just carries data, with case-by-case decisions on where to put the rules.

Reply:
> Alternatively we could start treating it as a pure gossip channel that just carries data, with case-by-case decisions on where to put the rules.

Yes, this pr is shifting towards this direction. It abstracts from how/if the certificate should be persisted.

Reply:
I think an interface based on shared observable memory is a guideline we can certainly explore, ie. instead of us providing runners to poll the stores and place the data into watches, we can eschew as much as we can of such runners on this side and make it clear that the responsibility to maintain the watchable input state and to subscribe to the watchable output state lies with the zksync-era side. We have a lot of code that we own there, so we can maintain these runners ourselves, but hopefully the choice of where the code should live would be obvious in most cases.